### ISSUE TITLE
Exception Handling considered critical

### ISSUE THREAD

**Issue by webyrd** [NCATSTranslator/ReasonerAPI#132]
In a production system it isn't unusual for 80% of the implementation of critical functions or methods to be exception and error handling code.

If I understand correctly, there is currently no standard way to report exceptional conditions, other than an error code mechanism or the logging mechanism.

This sounds like a giant hole, for building any production system, especially involving as many components or teams as we have.

Is there some minimal exception reporting mechanism we want to standardize on?

**Comment by edeutsch** [comment]
What we do in ARAX is use the `extended` schema properties Message.message_code and Message.code_description:
```
        message_code:
          type: string
          example: OK
          description: >-
            Set to OK for success, or some other short string to indicate and
            error (e.g., KGUnavailable, TermNotFound, etc.)
        code_description:
          type: string
          example: 9 results found
          description: >-
            Extended description denoting the success or mode of failure in the
            generation of the message
```

Our system is currently returning things like this:
```
message_code	code_description
OK	9 results found
QueryGraphInterpreterUnsupportedGraph	QueryGraphInterpreter finds more nodes than supported in this QueryGraph
QueryGraphZeroNodes	QueryGraph has 0 nodes. At least 1 node is required
QueryGraphTooManyEdges	QueryGraph may not have edges if there is only one node
QueryGraphEdgeWithNoId	QueryGraph has a edge with no id. This is not permitted
NoResults	No paths were found in {kp_to_use} satisfying this query graph
InvalidQuery	Cannot expand an edge for which neither end has any curies. (Could not find curies to use from a prior expand step, and neither qnode has a curie specified.)
```
We do not have a dictionary of message_codes, but that seems possible and sensible.

**Comment by patrickkwang** [comment]
@webyrd What features are we looking for from this exception reporting? HTTP status codes seem the natural mechanism for reporting errors, and I had been assuming that everyone would use them. But maybe you're looking for standardization on finer details than `400 Bad Request` or `500 Internal Server Error`?

**Comment by webyrd** [comment]
@patrickkwang I'm thinking of something much more fine grained, that would allow for *recovery* from exceptional states.  @edeutsch has some nice examples in his response above, such as TermNotFound or KGUnavailable.

> We do not have a dictionary of message_codes, but that seems possible and sensible.

That sounds like a reasonable start.

**Comment by patrickkwang** [comment]
Cool. We could definitely come up with some standards for the body of 400 responses, to describe malformed query graphs, unrecognized identifiers, and the like. I'm not sure how many of these will be autonomously recoverable, but it won't hurt, and can definitely help developers.

**Comment by vdancik** [comment]
We should classify exceptions as errors and warnings. For errors it is appropriate to return non-200 error codes along with a simple JSON object containing details of what went wrong; here is an example from an internal MolePro API:
```
{
    "status": 403,
    "title": "Not Found",
    "detail": "Collection bhtcbcXcu6 not found",
    "type": "about:blank"
}
```
For warnings we can include a warning array (probably in Result object) that could look like this:
```
    "warnings": [
        {
            "warning": "something wrong",
            "edge_id": "e01",
            "node_id": null,
            "detail": "Server providing extra information timed out"
        }
    ]
```

**Comment by vdancik** [comment]
Error message used by MolePro (and default python-flask-server error message)

```
    error_msg:
      type: object
      properties:
        status:
          type: integer
        title:
          type: string
        detail:
          type: string
        type:
          type: string
```

**Comment by patrickkwang** [comment]
Addressed by #158.
